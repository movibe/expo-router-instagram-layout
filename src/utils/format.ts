export const firstName = (name: string) => name?.split(' ')?.[0]

export const cleanProperties = (properties: Record<string, any>) => {
  const clean = {} as Record<string, any>
  Object.keys(properties).forEach(key => {
    if (properties[key] !== undefined) {
      clean[key] = properties[key]
    }
    if (typeof properties[key] === 'object') {
      clean[key] = JSON.stringify(properties[key])
    }
  })
  return clean
}

export const digitsOnly = (string: string) => string.replace(/\D/g, '')

export const replaceUrlValues = (url: string, queryParams: { [key: string]: string }): string => {
  for (const key in queryParams) {
    const pattern = `{${key}}`
    url = decodeURI(url).replace(pattern, queryParams[key])
  }
  return encodeURI(url)
}

export const replaceValues = (text: string, queryParams: { [key: string]: string }): string => {
  for (const key in queryParams) {
    const pattern = `{${key}}`
    text = text.replace(pattern, queryParams[key])
  }
  return text
}

export const currencyFormat = (amount: number): string => {
  return new Intl.NumberFormat('pt-BR', {
    currency: 'BRL',
    style: 'currency'
  }).format(amount / 100)
}

export const printPhone = (phone: string): string => {
  if (!phone || phone.length < 11) throw new Error('invalid phone length')
  return `(${phone.slice(-11, -9)}) ${phone.slice(-9, -4)}-${phone.slice(-4)}`
}

export const formatCpf = (cpf: string): string =>
  cpf.replace(/(\d{3})(\d{3})(\d{3})(\d{2})/, '$1.$2.$3-$4')

export const formatCep = (cep: string): string => cep.replace(/(\d{5})(\d{3})/, '$1-$2')

export function compareVersions(
  v1: string = '2.0.0',
  comparator: string = '>',
  v2: string = '1.0.0'
): boolean {
  const comparators = ['==', '===', '<', '<=', '>', '>=', '!=', '!==']

  if (!comparators.includes(comparator)) {
    throw new Error('Invalid comparator')
  }

  const v1parts = v1.split('.').map(part => parseInt(part, 10))
  const v2parts = v2.split('.').map(part => parseInt(part, 10))

  for (let i = 0; i < Math.max(v1parts.length, v2parts.length); i++) {
    const part1 = v1parts[i] || 0
    const part2 = v2parts[i] || 0

    if (part1 !== part2) {
      switch (comparator) {
        case '==':
        case '===':
          return false
        case '<':
          return part1 < part2
        case '<=':
          return part1 <= part2
        case '>':
          return part1 > part2
        case '>=':
          return part1 >= part2
        case '!=':
        case '!==':
          return true
      }
    }
  }

  switch (comparator) {
    case '==':
    case '===':
    case '>=':
    case '<=':
      return true
    case '<':
      return false
    case '>':
      return true
    case '!=':
    case '!==':
      return false
    default:
      return false
  }
}

export const parseStringToType = (value: string | boolean): boolean | string => {
  if (typeof value === 'boolean' || typeof value === 'number') return value

  const v = typeof value === 'string' ? value.toLowerCase().toString() : value
  if (v === 'true' || v === 'false') return v === 'true'

  return value
}
