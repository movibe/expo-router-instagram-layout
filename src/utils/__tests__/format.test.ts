import * as sut from '../format'

test('should format string', () => {
  const result = sut.firstName('John Wick')
  expect(result).toBe('John')
})

test('should clean properties', () => {
  const result = sut.cleanProperties({
    age: undefined,
    distance: 10,
    name: 'John Wick',
    properties: {
      age: undefined,
      distance: 10,
      name: 'John Wick'
    }
  })
  expect(result).toEqual({
    distance: 10,
    name: 'John Wick',
    properties: '{"distance":10,"name":"John Wick"}'
  })
})

test('should return digits only', () => {
  const result = sut.digitsOnly('John Wick 123')
  expect(result).toBe('123')
  expect(result).toMatchSnapshot()
})

describe('currencyFormat', () => {
  it('should format currency', () => {
    const result = sut.currencyFormat(1000)
    expect(result).toBe('R$ 10,00')
    expect(result).toMatchSnapshot()
  })

  it('should format currency R$ 100,00', () => {
    const result = sut.currencyFormat(10000)
    expect(result).toBe('R$ 100,00')
    expect(result).toMatchSnapshot()
  })

  it('should format currency R$ 1.000,00', () => {
    const result = sut.currencyFormat(100000)
    expect(result).toBe('R$ 1.000,00')
    expect(result).toMatchSnapshot()
  })
})

describe('printPhone', () => {
  it('should print phone', () => {
    const result = sut.printPhone('5511999999999')
    expect(result).toBe('(11) 99999-9999')
    expect(result).toMatchSnapshot()
  })

  it('should throw error', () => {
    const error = new Error('invalid phone length')
    expect(() => sut.printPhone('99999')).toThrow(error)
  })
})

describe('formatCpf', () => {
  it('should format cpf', () => {
    const result = sut.formatCpf('12345678909')
    expect(result).toBe('123.456.789-09')
    expect(result).toMatchSnapshot()
  })
})

describe('formatCep', () => {
  it('should format cep', () => {
    const result = sut.formatCep('12345678')
    expect(result).toBe('12345-678')
    expect(result).toMatchSnapshot()
  })
})

describe('compareVersions', () => {
  it('should return true for greater version - 2.0.0 > 1.0.0', () => {
    const result = sut.compareVersions('2.0.0', '>', '1.0.0')
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should return true for greater version - 2.0.0 >= 1.0.0', () => {
    const result = sut.compareVersions('2.0.0', '>=', '2.0.0')
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should return true for greater version - 1.0.0 < 2.0.0', () => {
    const result = sut.compareVersions('1.0.0', '<', '2.0.0')
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should return true for greater version - 1.0.0 <= 2.0.0', () => {
    const result = sut.compareVersions('2.0.0', '<=', '2.0.0')
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should return false for lower version - 1.0.0 > 2.0.0', () => {
    const result = sut.compareVersions('1.0.0', '>', '2.0.0')
    expect(result).toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should return false for lower version - 1.0.0 >= 2.0.0', () => {
    const result = sut.compareVersions('1.0.0', '>=', '2.0.0')
    expect(result).toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should throw for invalid operator - 1.0.0 a 2.0.0', () => {
    const error = new Error('Invalid comparator')
    expect(() => sut.compareVersions('1.0.0', 'a', '2.0.0')).toThrow(error)
  })
})

describe('parseStringToType', () => {
  it('should parse string to type TRUE', () => {
    const result = sut.parseStringToType('TRUE')
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })
  it('should parse string to type true', () => {
    const result = sut.parseStringToType(true)
    expect(result).not.toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should parse string to type FALSE', () => {
    const result = sut.parseStringToType('FALSE')
    expect(result).toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should parse string to type false', () => {
    const result = sut.parseStringToType(false)
    expect(result).toBeFalsy()
    expect(result).toMatchSnapshot()
  })

  it('should parse string to type 1', () => {
    const result = sut.parseStringToType('1')
    expect(result).toBe('1')
    expect(result).toMatchSnapshot()
  })
})
