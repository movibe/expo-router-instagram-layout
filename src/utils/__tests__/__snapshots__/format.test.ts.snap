// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareVersions should return false for lower version - 1.0.0 > 2.0.0 1`] = `false`;

exports[`compareVersions should return false for lower version - 1.0.0 >= 2.0.0 1`] = `false`;

exports[`compareVersions should return true for greater version - 1.0.0 < 2.0.0 1`] = `true`;

exports[`compareVersions should return true for greater version - 1.0.0 <= 2.0.0 1`] = `true`;

exports[`compareVersions should return true for greater version - 2.0.0 > 1.0.0 1`] = `true`;

exports[`compareVersions should return true for greater version - 2.0.0 >= 1.0.0 1`] = `true`;

exports[`currencyFormat should format currency 1`] = `"R$ 10,00"`;

exports[`currencyFormat should format currency R$ 1.000,00 1`] = `"R$ 1.000,00"`;

exports[`currencyFormat should format currency R$ 100,00 1`] = `"R$ 100,00"`;

exports[`formatCep should format cep 1`] = `"12345-678"`;

exports[`formatCpf should format cpf 1`] = `"123.456.789-09"`;

exports[`parseStringToType should parse string to type 1 1`] = `"1"`;

exports[`parseStringToType should parse string to type FALSE 1`] = `false`;

exports[`parseStringToType should parse string to type TRUE 1`] = `true`;

exports[`parseStringToType should parse string to type false 1`] = `false`;

exports[`parseStringToType should parse string to type true 1`] = `true`;

exports[`printPhone should print phone 1`] = `"(11) 99999-9999"`;

exports[`should return digits only 1`] = `"123"`;
